2022/Q4/artifacts/PoS-pseudocode/PoS-model.md

```diff
@@ -301,34 +303,43 @@ func unbond(validator_address, delegator_address, total_amount)
-//This function is called by transactions tx_withdraw_unbonds_validator and tx_withdraw_unbonds_delegator
+// This function is called by transactions tx_withdraw_unbonds_validator and tx_withdraw_unbonds_delegator.
+// Withdraws all tokens from PoS for all unbonds between the delegator adn validator that are fully matured as of the current epoch. Applies slashes to the unbonds, transferring all slashed tokens to the slash pool address and all unslashed tokens back to the source delegator.
 func withdraw(validator_address, delegator_address)
 {
   var delunbonds = {<start,end,amount> | amount = unbonds[delegator_address][validator_address].deltas[(start, end)] > 0 && end <= cur_epoch }
-  //substract any pending slash before withdrawing
   forall (<start,end,amount> in selfunbonds) do
-    // set of slashes that happened while the bond associated with the unbond was contributing to the validator's stake
-    var set_slashes = {s | s in slashes[validator_address] && start <= slash.epoch && slash.epoch < end - unbonding_length }
+    // Apply slashes that happened while the bond associated with the unbond was contributing to the validator's stake
+    var set_slashes = {s | s in slashes[validator_address] && start <= slash.epoch && slash.epoch < end - unbonding_length - cubic_slash_window_width}
     var amount_after_slashing = compute_amount_after_slashing(set_slashes, amount)
     balance[delegator_address] += amount_after_slashing
-    balance[pos] -= amount_after_slashing
-    //remove unbond
+    balance[slash_pool] += amount - amount_after_slashing
+    balance[pos] -= amount
+    // Remove the unbond
     unbonds[delegator_address][validator_address].deltas[(start,end)] = 0
 }
 ```
 
 ```go
 compute_amount_after_slashing(set_slashes, amount) {
+  // First, group the slashes by epoch and a total rate
+  var slash_rates = {}
+  forall (s in set_slashes) do
+    slash_rates[s.epoch] = min{1.0, slash_rate[s.epoch] + s.rate}
+  
   var computed_amounts = {}
-    var updated_amount = amount
-    forall (slash in set_slashes in slash.epoch order) do
-      //Update amount with slashes that happened more than `unbonding_length` before this slash
-      forall (slashed_amount in computed_amounts s.t. slashed_amount.epoch + unbonding_length < slash.epoch) do
-        updated_amount -= slashed_amount.amount
-        computed_amounts = computed_amounts \ {slashed_amount}
-      computed_amounts = computed_amounts \union {SlashedAmount{epoch: slash.epoch, amount: updated_amount*slash.rate}}
+  var updated_amount = amount
 
-    return updated_amount - sum({computed_amount.amount | computed_amount in computed_amounts})
+  // Now apply the slashes
+  forall (slash in slash_rates in slash.epoch order) do
+    // Update amount with slashes that happened more than `unbonding_length` before this slash
+    forall (slashed_amount in computed_amounts s.t. slashed_amount.epoch + unbonding_length < slash.epoch) do
```

Comment:
Should we now add the window width here: `s.t. slashed_amount.epoch + unbonding_length + cubic_slash_window_width < slash.epoch`?

Reply:
I think so

Reply:
I think so for all the cases.